integrationtest/mapper/pojo-base/src/test/java/org/hibernate/search/integrationtest/mapper/pojo/massindexing/MassIndexingBatchingThreadContextIT.java

```diff
@@ -0,0 +1,147 @@
+import org.hibernate.search.util.common.impl.Futures;
+import org.hibernate.search.util.impl.integrationtest.common.rule.BackendMock;
+import org.hibernate.search.util.impl.integrationtest.mapper.pojo.standalone.StandalonePojoMappingSetupHelper;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class MassIndexingBatchingThreadContextIT {
+
+	@Rule
+	public final BackendMock backendMock = new BackendMock();
+
+	@Rule
+	public final StandalonePojoMappingSetupHelper setupHelper
+			= StandalonePojoMappingSetupHelper.withBackendMock( MethodHandles.lookup(), backendMock );
+
+	private SearchMapping mapping;
+
+	private final StubLoadingContext loadingContext = new StubLoadingContext();
+
+	@Before
+	public void setup() {
+		backendMock.expectAnySchema( Entity.INDEX );
+
+		mapping = setupHelper.start()
+				.withConfiguration( b -> {
+					b.addEntityType( Entity.class, c -> c
+							.massLoadingStrategy( new StubMassLoadingStrategy<>( Entity.PERSISTENCE_KEY ) ) );
+				} )
+				.setup( Entity.class );
+
+		backendMock.verifyExpectationsMet();
+
+		initData();
+	}
+
+	@Test
+	public void entityWithPrimitiveId() throws InterruptedException {
+		try ( SearchSession searchSession = mapping.createSession() ) {
+			AtomicInteger before = new AtomicInteger( 0 );
+			AtomicInteger after = new AtomicInteger( 0 );
+
+			MassIndexer indexer = searchSession.massIndexer()
+					// Simulate passing information to connect to a DB, ...
+					.context( StubLoadingContext.class, loadingContext )
+					.mergeSegmentsOnFinish( true )
+					.typesToIndexInParallel( 1 )
+					.threadsToLoadObjects( 1 )
+					.batchingThreadContext( new BatchingThreadContext() {
+						@Override
+						public void beforeExecution(Thread t, Runnable r) {
+							before.incrementAndGet();
+						}
+
+						@Override
+						public void afterExecution(Runnable r, Throwable t) {
+							after.incrementAndGet();
+						}
+					} );
+
+			// add operations on indexes can follow any random order,
+			// since they are executed by different threads
+			backendMock.expectWorks(
+							Entity.INDEX, DocumentCommitStrategy.NONE, DocumentRefreshStrategy.NONE
+					)
+					.add( "1", b -> {
+					} );
+
+			// purgeAtStart and mergeSegmentsAfterPurge are enabled by default,
+			// so we expect 1 purge, 1 mergeSegments and 1 flush calls in this order:
+			backendMock.expectIndexScaleWorks( Entity.INDEX, searchSession.tenantIdentifier() )
+					.purge()
+					.mergeSegments()
+					.mergeSegments()
+					.flush()
+					.refresh();
+
+			Futures.unwrappedExceptionGet( indexer.start().toCompletableFuture() );
+
+			assertThat( before ).hasValue(
+					0 // Hibernate Search - Mass indexing - Coordinator - 0
+							+ 1 // Hibernate Search - Mass indexing - Workspace - 0
+							+ 1 // Hibernate Search - Mass indexing - Entity - Entity loading - 0
+							+ 1 // Hibernate Search - Mass indexing - Entity - ID loading - 0
```

Comment:
Maybe push the thread names to a (threasafe) `List` , so that you can check that explicitly and don't have to guess the threads involved here?